samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java

```diff
@@ -16,85 +16,327 @@
  * under the License.
  */
+
 package org.apache.samza.operators.windows;
 
+import org.apache.samza.annotation.InterfaceStability;
 import org.apache.samza.operators.data.MessageEnvelope;
+import org.apache.samza.operators.triggers.TimeTrigger;
+import org.apache.samza.operators.triggers.Trigger;
+import org.apache.samza.operators.triggers.Triggers;
 
+import java.time.Duration;
 import java.util.Collection;
+import java.util.function.BiFunction;
 import java.util.function.Function;
 
-
 /**
- * This class defines a collection of {@link Window} functions. The public classes and methods here are intended to be
- * used by the user (i.e. programmers) to create {@link Window} function directly.
+ * Factory methods for creating a {@link Window}. Provides APIs for specifying different types of {@link Window}s.
+ * A {@link Window} can have a specified keyFunction that is used to group {@link MessageEnvelope}s in the {@link Window}.
  *
+ * <p> The following types of {@link Window}s are supported.
+ * <ul>
+ *  <li>Tumbling Windows: A tumbling window defines a series of non-overlapping, fixed size, contiguous time intervals.
+ *  <li>Session Windows: A session window groups a {@link org.apache.samza.operators.MessageStream} into sessions.
+ *  A <i>session</i> captures some period of activity over a {@link org.apache.samza.operators.MessageStream}.
+ *  The boundary for a session is defined by a {@code sessionGap}. All {@link MessageEnvelope}s that that arrive within
+ *  the gap are grouped into the same session.
+ *  <li>Global Windows: A global window defines a single infinite window over the entire {@link org.apache.samza.operators.MessageStream}.
+ *  An early trigger must be specified when defining a global window.
+ * </ul>
  */
+@InterfaceStability.Unstable
 public final class Windows {
 
+  private Windows() {
+
+  }
+
   /**
-   * private constructor to prevent instantiation
+   * Creates a {@link Window} that groups incoming {@link MessageEnvelope}s using the provided keyFn into
+   * fixed-size, non-overlapping processing time based windows and applies the provided aggregate function to them.
+   *
+   * <p>The example below computes the maximum value per-key over fixed size 10 second windows.
+   *
+   * <pre> {@code
+   * MessageStream<UserClick> stream = ...;
+   * Function<UserClick, String> keyFn = ...;
+   * BiFunction<UserClick, Integer, Integer> maxAggregator = (m, c)-> Math.max(parseInt(m), c);
+   * MessageStream<WindowOutput<WindowKey<String>, Integer>> windowedStream = stream.window(
+   * Windows.keyedTumblingWindow(keyFn, Duration.seconds(10), maxAggregator));
+   * }
+   * </pre>
+   *
+   * @param keyFn the function to extract the key from a {@link MessageEnvelope}
+   * @param interval the duration in processing time.
+   * @param aggregateFunction the function to aggregate window results
+   * @param <M> the type of the input {@link MessageEnvelope}
+   * @param <WV> the type of the {@link WindowOutput} output value
```

Comment:
the type of the value in WindowOutput?

How about renaming WindowOutput to WindowPane for consistency?